app/src/main/java/com/example/dungeontest/MainScreen.kt

```
@@ -61,25 +66,38 @@ fun MainScreen(
     val context = LocalContext.current
     val preferences = SettingsStorage(context)
     val viewModel = viewModel<MapViewModel>()
-    val maps = viewModel.allMaps.observeAsState()
-    var photoUri by remember { mutableStateOf<Uri?>(null) }
-    val snackbarHostState = remember { SnackbarHostState() }
-    val cameraLauncher = rememberLauncherForActivityResult(
-        contract = ActivityResultContracts.TakePicture()
-    ) { success ->
-        if (success) {
-            photoUri?.let {
-                Log.v("MainScreen", "Photo URI: $it")
-                /* Encode the URI to base64 because it's a file path. Navigation uses / */
+    
+    val options = GmsDocumentScannerOptions.Builder()
+        .setGalleryImportAllowed(true)
+        .setPageLimit(1)
+        .setResultFormats(RESULT_FORMAT_JPEG)
+        .setScannerMode(SCANNER_MODE_FULL)
+        .build()
 
-                 val base64EncodedUri = Base64.encodeToString(it.toString().toByteArray(Charsets.UTF_8), Base64.NO_WRAP)
-                /* navcontroller is passed from the NavGraph setup in MainActivity */
-                navController.navigate("NamingScreen/$base64EncodedUri") {
-                    popUpTo("NamingScreen") { inclusive = true }
+    val scanner = GmsDocumentScanning.getClient(options)
+    var photoUri by remember {
+        mutableStateOf<Uri?>(null)
+    }
+    val scannerLauncher = rememberLauncherForActivityResult(
+        contract = ActivityResultContracts.StartIntentSenderForResult(),
+        onResult = {
+            if (it.resultCode == RESULT_OK) {
+                val result = GmsDocumentScanningResult.fromActivityResultIntent(it.data)
+                photoUri = result?.pages?.get(0)?.imageUri
+                photoUri?.let {
```

to avoid 'silent failure', imo it would be good to use an if-else here, i.e.
```
if(photoUri != null) {
                    val base64EncodedUri = Base64.encodeToString(photoUri.toString().toByteArray(Charsets.UTF_8), Base64.NO_WRAP)
                    navController.navigate("NamingScreen/$base64EncodedUri") {
                        popUpTo("NamingScreen") { inclusive = true }
                    }
} else {
                    //todo add logger call here, preferably printing a jsonified version of the result object or something like that
}
```

---

sorry, initial code suggestion was using typescript syntax (null being falsy)

---

Thank you, addressed